flang/lib/Lower/CharacterExpr.cpp

```diff
@@ -180,13 +180,21 @@ static mlir::Type getSingletonCharType(mlir::MLIRContext *ctxt, int kind) {
 
 mlir::Value
 Fortran::lower::CharacterExprHelper::createEmbox(const fir::CharBoxValue &box) {
-  // BoxChar require a reference. Base CharBoxValue of CharArrayBoxValue
-  // are ok here (do not require a scalar type)
+  // Base CharBoxValue of CharArrayBoxValue are ok here (do not require a scalar
+  // type)
   auto charTy = recoverCharacterType(box.getBuffer().getType());
   auto boxCharType =
       fir::BoxCharType::get(builder.getContext(), charTy.getFKind());
   auto refType = fir::ReferenceType::get(boxCharType.getEleTy());
-  auto buff = builder.createConvert(loc, refType, box.getBuffer());
+  mlir::Value buff = box.getBuffer();
+  // fir.boxchar requires a memory reference. Allocate temp if the character is
+  // not in memory.
+  if (!fir::isa_ref_type(buff.getType())) {
```

Comment:
Should a CharBoxValue always have a base value that is a memory reference?

Reply:
I think it would be simpler and safer. We used to load constant length characters, but that is not true anymore, at least in `genLoad`. CHAR intrinsic lowering may be the only place that still produce a character that is not in memory, so I think it is worth considering enforcing having characters always in memory. I won't try this in this PR though.